cumulus/pallets/collator-selection/src/lib.rs

```diff
@@ -554,6 +630,156 @@ pub mod pallet {
 			Self::deposit_event(Event::InvulnerableRemoved { account_id: who });
 			Ok(())
 		}
+
+		/// Update the candidacy bond of collator candidate `origin` to a new amount `new_deposit`.
+		///
+		/// Setting a `new_deposit` that is lower than the current deposit while `origin` is
+		/// occupying a top-`DesiredCandidates` slot is not allowed.
+		///
+		/// This call will fail if `origin` is not a collator candidate, the updated bond is lower
+		/// than the minimum candidacy bond, and/or the amount cannot be reserved.
+		#[pallet::call_index(7)]
+		#[pallet::weight(T::WeightInfo::update_bond(T::MaxCandidates::get()))]
+		pub fn update_bond(
+			origin: OriginFor<T>,
+			new_deposit: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			ensure!(new_deposit >= <CandidacyBond<T>>::get(), Error::<T>::DepositTooLow);
+			// The function below will try to mutate the `CandidateList` entry for the caller to
+			// update their deposit to the new value of `new_deposit`. The return value is the
+			// position of the entry in the list, used for weight calculation.
+			let length =
+				<CandidateList<T>>::try_mutate(|candidates| -> Result<usize, DispatchError> {
+					let idx = candidates
+						.iter()
+						.position(|candidate_info| candidate_info.who == who)
+						.ok_or_else(|| Error::<T>::NotCandidate)?;
+					let candidate_count = candidates.len();
+					// Remove the candidate from the list.
+					let mut info = candidates.remove(idx);
+					let old_deposit = info.deposit;
+					if new_deposit > old_deposit {
+						T::Currency::reserve(&who, new_deposit - old_deposit)?;
+					} else if new_deposit < old_deposit {
+						// Casting `u32` to `usize` should be safe on all machines running this.
+						ensure!(
+							idx.saturating_add(<DesiredCandidates<T>>::get() as usize) <
+								candidate_count,
+							Error::<T>::InvalidUnreserve
+						);
+						T::Currency::unreserve(&who, old_deposit - new_deposit);
+					} else {
+						return Err(Error::<T>::IdenticalDeposit.into())
+					}
+
+					// Update the deposit and insert the candidate in the correct spot in the list.
+					info.deposit = new_deposit;
+					let new_pos = candidates
+						.iter()
+						.position(|candidate| candidate.deposit >= new_deposit)
+						.unwrap_or_else(|| candidates.len());
+					candidates
+						.try_insert(new_pos, info)
+						.expect("candidate count previously decremented; qed");
```

Comment:
better an error here instead panic

Reply:
The `info` we are `try_insert`ing was just removed from the `BoundedVec`, it is impossible to fail.

Reply:
yes, but you never now how the code will evolve

Reply:
Removed the `expect`, now it's an error.